README.md
[![Yii2](https://img.shields.io/badge/required-Yii2_v2.0.40-blue.svg)](https://packagist.org/packages/yiisoft/yii2)
[![Downloads](https://img.shields.io/packagist/dt/wdmg/yii2-search.svg)](https://packagist.org/packages/wdmg/yii2-search)
[![Packagist Version](https://img.shields.io/packagist/v/wdmg/yii2-search.svg)](https://packagist.org/packages/wdmg/yii2-search)
![Progress](https://img.shields.io/badge/progress-ready_to_use-green.svg)
[![GitHub license](https://img.shields.io/github/license/wdmg/yii2-search.svg)](https://github.com/wdmg/yii2-search/blob/master/LICENSE)

<img src="./docs/images/yii2-search.png" width="100%" alt="Yii2 Search module" />

# Yii2 Search
Search module for Yii2.

The module implements indexed search using morphology (phpMorphy) or Porter's stemmer algorithm (LinguaStem). A live full-text search by data model is also provided.

This module is an integral part of the [Butterfly.СMS](https://butterflycms.com/) content management system, but can also be used as an standalone extension.

Copyrights (c) 2019-2023 [W.D.M.Group, Ukraine](https://wdmg.com.ua/)

# Requirements 
* PHP 5.6 or higher
* Yii2 v.2.0.40 and newest
* [Yii2 Base](https://github.com/wdmg/yii2-base) module (required)
* [phpMorphy](https://github.com/wdmg/phpmorphy) library
* [LinguaStem](https://github.com/wdmg/lingua-stem) library
* [Yii2 Helpers](https://github.com/wdmg/yii2-helpers)

# Installation
To install the module, run the following command in the console:

`$ composer require "wdmg/yii2-search"`

After configure db connection, run the following command in the console:

`$ php yii search/init`

And select the operation you want to perform:
  1) Apply all module migrations
  2) Revert all module migrations

# Migrations
In any case, you can execute the migration and create the initial data, run the following command in the console:

`$ php yii migrate --migrationPath=@vendor/wdmg/yii2-search/migrations`

# Configure
To add a module to the project, add the following data in your configuration file:

    'modules' => [
        ...
        'search' => [ // list of supported models for live search or/and indexation
            'class' => 'wdmg\search\Module',
            'routePrefix' => 'admin',
            'supportModels' => [
               'news' => [
                   'class' => 'wdmg\news\models\News',
                   'indexing' => [
                       'on_insert' => true,
                       'on_update' => true,
                       'on_delete' => true
                   ],
                   'options' => [
                       'title' => 'title', // attr name (string) of model or function($model)
                       'url' => 'url', // attr name (string) of model or function($model)
                       'fields' => [
                           'title',
                           'keywords',
                           'description',
                           'content'
                       ],
                       'conditions' => [
                           'status' => 1
                       ]
                   ]
               ],
               ...
           ],
           'cacheExpire' = 86400, // live search cache lifetime, `0` - for not use cache
           'indexingOptions' = [ // indexation options
               'processing' => 'phpMorphy', //  Set `phpMorphy` or `LinguaStem`
               'language' => 'ru-RU', // Support 'ru-RU', 'uk-UA', 'de-DE'
               'analyze_by' => 'relevance',
               'max_execution_time' => 0, // max execution time in sec. for indexing process
               'memory_limit' => null, // max operating memory in Mb for indexing process
               'max_words' => 50,
           ],
           'analyzerOptions' = [ // text analyzer options, see \wdmg\helpers\TextAnalyzer
               'min_length' => 3,
               'stop_words' => [],
               'weights' => []
           ],
           'snippetOptions' = [ // build search snippet options
               'max_words_before' => 6,
               'max_words_after' => 4,
               'bolder_tag' => 'strong',
               'max_length' => 255,
               'delimiter' => '…'
           ],
           'searchAccuracy' = 90, // search accuracy
        ],
        ...
    ],


# Routing
Use the `Module::dashboardNavItems()` method of the module to generate a navigation items list for NavBar, like this:

    <?php
        echo Nav::widget([
        'options' => ['class' => 'navbar-nav navbar-right'],
            'label' => 'Modules',
            'items' => [
                Yii::$app->getModule('search')->dashboardNavItems(),
                ...
            ]
        ]);
    ?>

# Status and version [ready to use]
* v.2.0.0 - Update copyrights, fix nav menu
* v.1.1.4 - Fixed mysql syntax error: 1055 for MySQL >= 5.7
* v.1.1.3 - Update dependencies

CHANGELOG.md
Changelog
=========

## 2.0.0 (2023-06-25)
 * Update copyrights, fix nav menu
 * Show modal call fixed
 
## 1.1.4 (2021-01-17)
 * Fixed mysql syntax error: 1055 for MySQL >= 5.7
    
## 1.1.3 (2020-12-24)
 * Update dependencies
 
## 1.1.2 (2020-06-28)
 * Update dependencies, README.md

## 1.1.1 (2020-05-08)
 * Added flags, fixed indexing by models events
 * Added multi-language support
 * Indexing on create/update events of support models
 
## 1.1.0 (2020-05-07)
 * Update dependencies
 * Added LinguaStem() (support for Porter Stemmer)
 
## 1.0.10 (2020-04-28)
 * Update README.md and dependencies
 
## 1.0.9 (2020-04-14)
 * Checking module is loaded
 * Migrations fixed
 
## 1.0.8 (2020-03-25)
 * Up to date dependencies
 * Fix migrations
 
## 1.0.7 (2020-03-07)
 * Added support for Blog module
 
## 1.0.6 (2020-02-23)
 * Added no caching parameter for search method
 
## 1.0.5 (2020-02-16)
 * URL for snippets as closure object
 * Added ignored patterns
 
## 1.0.4 (2020-02-12)
 * Added drop search index and rebuild from dashboard
 
## 1.0.3 (2020-02-11)
 * Added drop search index and rebuild from console
 
## 1.0.2 (2020-02-10)
 * Added README.md, complete Search and LiveSearch interface
 
## 1.0.1 (2020-02-03)
 * Added Search model, indexing functionality
 * Add snippets for indexing results
 * Added other models and fixing LiveSearch
 * Added Search model, indexing functionality
 
## 1.0.0 (2020-02-03)
 * Added Base module class, migrations


Module.php
<?php

namespace wdmg\search;

/**
 * Yii2 Search
 *
 * @category        Module
 * @version         2.0.0
 * @author          Alexsander Vyshnyvetskyy <alex.vyshnyvetskyy@gmail.com>
 * @link            https://github.com/wdmg/yii2-search
 * @copyright       Copyright (c) 2023 W.D.M.Group, Ukraine
 * @license         https://opensource.org/licenses/MIT Massachusetts Institute of Technology (MIT) License
 *
 */

use wdmg\search\models\Search;
use Yii;
use wdmg\base\BaseModule;
use yii\base\InvalidConfigException;
use yii\helpers\ArrayHelper;
use yii\helpers\Url;
use yii\helpers\Html;

/**
 * RSS-feed module definition class
 */
class Module extends BaseModule
{
    /**
     * {@inheritdoc}
     */
    public $controllerNamespace = 'wdmg\search\controllers';

    /**
     * {@inheritdoc}
     */
    public $defaultRoute = "list/index";

    /**
     * @var string, the name of module
     */
    public $name = "Search";

    /**
     * @var string, the description of module
     */
    public $description = "Site search";

    /**
     * @var array list of supported models for live search or/and indexation
     */
    public $supportModels = [
        'news' => [
            'class' => 'wdmg\news\models\News',
            'indexing' => [
                'on_insert' => true,
                'on_update' => true,
                'on_delete' => true
            ],
            'options' => [
                'title' => 'title',
                'url' => 'url',
                'fields' => [
                    'title',
                    'keywords',
                    'description',
                    'content'
                ],
                'conditions' => [
                    'status' => 1
                ]
            ]
        ],
        'blog' => [
            'class' => 'wdmg\blog\models\Posts',
            'indexing' => [
                'on_insert' => true,
                'on_update' => true,
                'on_delete' => true
            ],
            'options' => [
                'title' => 'title',
                'url' => 'url',
                'fields' => [
                    'title',
                    'keywords',
                    'description',
                    'content'
                ],
                'conditions' => [
                    'status' => 1
                ]
            ]
        ],
        'pages' => [
            'class' => 'wdmg\pages\models\Pages',
            'indexing' => [
                'on_insert' => true,
                'on_update' => true,
                'on_delete' => true
            ],
            'options' => [
                'title' => 'title',
                'url' => 'url',
                'fields' => [
                    'title',
                    'keywords',
                    'description',
                    'content'
                ],
                'conditions' => [
                    'status' => 1
                ]
            ]
        ],
    ];


    /**
     * @var int live search cache lifetime, `0` - for not use cache
     */
    public $cacheExpire = 86400; // 86400 1 day.

    /**
     * @var array indexation options
     */
    public $indexingOptions = [
        'processing' => 'phpmorphy', //  Set `phpmorphy` or `lingua-stem`
        'language' => 'ru-RU', // Support 'ru-RU', 'uk-UA', 'de-DE'
        'analyze_by' => 'relevance',
        'max_execution_time' => 0, // max execution time in sec. for indexing process
        'memory_limit' => null, // max operating memory in Mb for indexing process
        'max_words' => 50,
    ];

    /**
     * @var array text analyzer options, see \wdmg\helpers\TextAnalyzer
     */
    public $analyzerOptions = [
        'min_length' => 3,
        'stop_words' => [],
        'weights' => []
    ];

    /**
     * @var array build search snippet options
     */
    public $snippetOptions = [
        'max_words_before' => 6,
        'max_words_after' => 4,
        'bolder_tag' => 'strong',
        'max_length' => 255,
        'delimiter' => '…'
    ];

    /**
     * @var int search accuracy
     */
    public $searchAccuracy = 90;

    /**
     * @var string the module version
     */
    private $version = "2.0.0";

    /**
     * @var integer, priority of initialization
     */
    private $priority = 5;

    /**
     * {@inheritdoc}
     */
    public function init()
    {
        parent::init();

        // Set version of current module
        $this->setVersion($this->version);

        // Set priority of current module
        $this->setPriority($this->priority);

    }

    /**
     * {@inheritdoc}
     */
    public function dashboardNavItems($options = null)
    {
        $items = [
            'label' => $this->name,
            'url' => '#',
            'icon' => 'fa fa-fw fa-search',
            'active' => in_array(\Yii::$app->controller->module->id, [$this->id]),
            'items' => [
                [
                    'label' => Yii::t('app/modules/search', 'Search index'),
                    'url' => [$this->routePrefix . '/search/list/'],
                    'active' => (in_array(\Yii::$app->controller->module->id, ['search']) &&  Yii::$app->controller->id == 'list'),
                ],
                [
                    'label' => Yii::t('app/modules/search', 'Ignored list'),
                    'url' => [$this->routePrefix . '/search/ignored/'],
                    'active' => (in_array(\Yii::$app->controller->module->id, ['search']) &&  Yii::$app->controller->id == 'ignored'),
                ],
            ]
        ];


	    if (!is_null($options)) {

		    if (isset($options['count'])) {
			    $items['label'] .= '<span class="badge badge-default float-right">' . $options['count'] . '</span>';
			    unset($options['count']);
		    }

		    if (is_array($options))
			    $items = \wdmg\helpers\ArrayHelper::merge($items, $options);

	    }

	    return $items;
    }

    /**
     * {@inheritdoc}
     */
    public function bootstrap($app)
    {
        parent::bootstrap($app);

        if (isset(Yii::$app->params["search.supportModels"]))
            $this->supportModels = Yii::$app->params["search.supportModels"];

        if (isset(Yii::$app->params["search.cacheExpire"]))
            $this->cacheExpire = Yii::$app->params["search.cacheExpire"];

        if (!isset($this->supportModels))
            throw new InvalidConfigException("Required module property `supportModels` isn't set.");

        if (!isset($this->cacheExpire))
            throw new InvalidConfigException("Required module property `cacheExpire` isn't set.");

        if (!is_array($this->supportModels))
            throw new InvalidConfigException("Module property `supportModels` must be array.");

        if (!is_integer($this->cacheExpire))
            throw new InvalidConfigException("Module property `cacheExpire` must be integer.");


        // Attach to events of create/change/remove of models for search indexing process
        if (!$this->isConsole() && is_array($models = $this->supportModels)) {

            $search = new Search();

            foreach ($models as $context => $support) {

                if (isset($support['class']) && isset($support['options'])) {

                    $class = $support['class'];
                    $options = $support['options'];

                    // If class of model exist
                    if (class_exists($class) && isset($support['indexing'])) {

                        $model = new $class();
                        $indexing = $support['indexing'];

                        if ($indexing['on_insert']) {
                            \yii\base\Event::on($class, $model::EVENT_AFTER_INSERT, function ($event) use ($search, $context, $options) {

                                $locale = null;
                                if (isset($event->sender->locale))
                                    $locale = $event->sender->locale;

                                $result = $search->indexing($event->sender, $context, $options, 1, $locale);

                                if ($result == 1 || $result == 2) {
                                    Yii::$app->getSession()->setFlash(
                                        'success',
                                        Yii::t('app/modules/search', 'OK! Material successfully indexed and added to search.')
                                    );
                                } else if ($result == -1) {
                                    Yii::$app->getSession()->setFlash(
                                        'danger',
                                        Yii::t('app/modules/search', 'An error occurred while indexing the material.')
                                    );
                                }
                            });
                        }

                        if ($indexing['on_update']) {
                            \yii\base\Event::on($class, $model::EVENT_AFTER_UPDATE, function ($event) use ($search, $context, $options) {

                                $locale = null;
                                if (isset($event->sender->locale))
                                    $locale = $event->sender->locale;

                                $result = $search->indexing($event->sender, $context, $options, 2, $locale);

                                if ($result == 1 || $result == 2) {
                                    Yii::$app->getSession()->setFlash(
                                        'success',
                                        Yii::t('app/modules/search', 'OK! Material successfully indexed and added to search.')
                                    );
                                } else if ($result == -1) {
                                    Yii::$app->getSession()->setFlash(
                                        'danger',
                                        Yii::t('app/modules/search', 'An error occurred while indexing the material.')
                                    );
                                }

                            });
                        }

                        if ($indexing['on_delete']) {
                            \yii\base\Event::on($class, $model::EVENT_AFTER_DELETE, function ($event) use ($search, $context, $options) {

                                $result = $search->indexing($event->sender, $context, $options, 3);

                                if ($result == 1 || $result == 2) {
                                    Yii::$app->getSession()->setFlash(
                                        'success',
                                        Yii::t('app/modules/search', 'OK! Material successfully deleted from search.')
                                    );
                                } else if ($result == -1) {
                                    Yii::$app->getSession()->setFlash(
                                        'danger',
                                        Yii::t('app/modules/search', 'An error occurred while deleteing the material from search.')
                                    );
                                }
                            });
                        }
                    }
                }
            }
        }
    }

    public function generateSnippets($content = null, $keywords = null, $options = [], $withBase = true)
    {

        $snippets = [];
        if (!empty($content) && is_array($keywords)) {

            foreach ($keywords as $keyword) {
                $count_in = 0;

                $min_words_before = 2;
                $max_words_before = (isset($options['before'])) ? $options['before'] : 6;
                $words_before = $min_words_before . ',' . intval($max_words_before);

                $min_words_after = 2;
                $max_words_after = (isset($options['after'])) ? $options['after'] : 4;
                $words_after = $min_words_after . ',' . intval($max_words_after);

                preg_match_all('/(?:\w+){' . $words_before . '}\s+((?:\b' . $keyword . '*?)\w+)\s+(?:\w+){' . $words_after . '}/iu', $content, $matches, PREG_OFFSET_CAPTURE);
                foreach ($matches as $key => $match) {


                    foreach ($match as $part) {

                        if ($count_in >= 3)
                            continue;

                        if ($tag = $options['tag'])
                            $snippet = preg_replace('/(?:\b' . $keyword . '*?)\w+/iu', Html::tag($tag, '$0'), $part[0]);

                        if (!empty($snippet)) {
                            if (empty($snippets)) {
                                
                                if ($withBase)
                                    $snippets[$keyword][] = mb_convert_case(mb_substr(trim($snippet), 0, 1), MB_CASE_TITLE) . mb_substr(trim($snippet), 1);
                                else
                                    $snippets[] = mb_convert_case(mb_substr(trim($snippet), 0, 1), MB_CASE_TITLE) . mb_substr(trim($snippet), 1);

                            } else {

                                if ($withBase)
                                    $snippets[$keyword][] = trim($snippet);
                                else
                                    $snippets[] = trim($snippet);

                            }

                            $count_in++;
                        }

                    }
                }
            }
        }

        return $snippets;
    }
}
